docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+Another relevant effort is that of ipld selectors.  ipld selectors are interesting because they are an orthagonal approach to the Cluster DAG for succinctly organizing large pinsets.  We did not end up including them in the main proposal because adding large datasets necessitates importing data to a DAG anyway.  In this case, from our perspective, building a management DAG for pinning and directly transferring blocks is simpler than adding a portion of the DAG to the local ipfs node, executing a remote selector pin, blocking on ipfs to do the transfer, and then clearing resources from the local node.  However in the case that the dag is already on ipfs, pinning selectors for shards is a potential implementation strategy.  The selectors would essentially act as the shard-nodes of the Cluster DAG, selecting shards of the DAG that could fit on individual nodes and getting pinned by the assigned nodes.  At the very least this seems to require that ipld selectors can select based on sub-tree size.  It is currently unclear how well matched ipld selectors would be for this use case.
+
+# Workplan
+
+There is a fair amount that needs to be built, updated and serviced to make all of this work.  Here I am focusing on basic sharding, allowing support for files that wouldn't fit in an ipfs repo.  Support for fault tolerance, more clever sharding strategies and sharding large existing DAGs will come after this is nailed down.  
+
+* Build up a version of the `importers` (aka `dagger` aka `dex`) project that suits cluster's needs (if possible keep it general enough for general use).  In short we want a library that imports data to ipfs DAGs from arbitrary formats (tar, deb, unix files, etc) with arbitrary chunking and layout algorithms.  For our use case it is important that the interfaces support taking in a stream of data and returning a stream of blocks or DAG nodes.
+* Implement a cluster importer component.  It will need to provide endpoints that supporting streams as input and output.
+* Create a cluster add endpoint on the REST API that reads in file data on HTTP multipart requests.  Implement forwarding of data to the cluster importer component.  Put the output blocks into the local ipfs daemon and pin the root with cluster.  This will require figuring out the connection strategy between the importer component and the REST API and nailing down the APIs of these components.
```

Comment:
Would need rewrite:  chunking happens directly in the component using `dex`. Might need a simple sharding component already, even if it doesn't shard and just allocates blocks.